lib/matplotlib/colorbar.py

```diff
@@ -673,7 +718,32 @@ def add_lines(self, levels, colors, linewidths, erase=True):
             for each line.
         erase : bool, default: True
             Whether to remove any previously added lines.
+
+        Notes
+        -----
+        For back-compatibility, this method can also be called with the
+        signature ``colorbar.add_lines(contour_set, erase=True)``, in which
+        case *levels*, *colors*, and *linewidths* are taken from *contour_set*.
         """
+        try:
+            self, CS, erase = (
+                lambda self, CS, erase=True: (self, CS, erase))(
+                    self, *args, **kwargs)
+        except TypeError:
+            self, levels, colors, linewidths, erase = (
```

Comment:
This is new, and probably needs a test.  I don't actually understand what this lambda does and I find it quite unreadable.  Same with the lambda above.   You have used this lambda trick elsewhere, is it possible to encapsulate it as an `_api` function that has some semantics?   

Also, does this need an API change note?  I am not clear on what API is even being changed here, but this seems somewhat unrelated to merging these two classes?

Reply:
Actually this is completely related to the merging of the two classes (and doesn't need an API change: it's here strictly for backcompat).
Previously, ColorbarBase.add_lines() had signature `add_lines(self, levels, colors, linewidths, erase=True)`, whereas Colorbar.add_lines() had signature `add_lines(self, CS, erase=True)`.  Therefore, if we want to merge the two classes while maintaining backcompat, we need to support both calling conventions.  In order to do so, we make `add_lines()` take `*args, **kwargs`, and then parse them using the helper lambdas: if `self, *args, **kwargs` can be passed to `lambda self, levels, colors, linewidths, erase=True: <thing>`, then the user called it using the ColorbarBase signature; otherwise, the user called it using the Colorbar signature.  The helper lambdas then perform two tasks: 1) if the call has a valid parameter list, then assign the right values to levels, colors, etc., and 2), if the call has an *invalid* parameter list, generate a standard error message, using Python's standard machinery ("expected 4 args but got 2" or "got an unexpected arg 'foo'" or whatnot).
Perhaps that can go into an _api helper, but let's do that another time?  (There's only one other use of the pattern in the library now, AFAICT, in QuadMesh.)

Reply:
Also note, both of these forms are inconsistent with Axes.add_lines :( So there are at least 3 different signatures floating around

Reply:
I'm not really sure add_lines should be public, but deprecating it is not the point of this PR.

Reply:
OK, can you at least add a comment explaining what is going on?  Again, this trick is very unintuitive to me, and if we are going to have very hard-to-follow code in the codebase, it at least needs comments explaining what the magical wand does.

Reply:
Actually, perhaps https://github.com/matplotlib/matplotlib/pull/20420 can go in first and I'll rebase this on top of it...